src/lib.rs

```diff
@@ -6,3 +6,24 @@
 #![deny(missing_debug_implementations)]
 #![deny(missing_docs)]
 #![deny(unsafe_code)]
+
+mod address;
+pub mod bundle;
+mod circuit;
+pub mod keys;
+mod note;
+pub mod primitives;
+mod tree;
+pub mod value;
+
+pub use address::Address;
+pub use note::{EncryptedNote, Note, NoteCommitment, Nullifier};
+
+/// Chain-specific constants and constraints for Orchard.
+///
+/// The purpose of this trait is to encapsulate things like the human-readable prefixes
+/// for encoded addresses, or the range of allowable values for notes.
+pub trait Chain {
```

Comment:
I'm not 100% sold on this trait. It has the flavor of the kind of constraints that are ultimately proving relatively trivial invariants in the type system, and which fall apart at serialization boundaries. It currently adds a fair bit of noise to the crate, and just strikes me as the sort of thing I've regretted doing in the past. Will have to consider further. It particularly seems a little bit superfluous given that Rust doesn't have GADTs, so none of the static invariants that this would cause to hold can be accessed dynamically without throwing away the type safety this would give you anyway (ala https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/consensus.rs#L273).

One alternative that occurs to me: instead of defining concrete data structures that are parameterized on `C: Chain` one could instead define a set of traits that represent the capabilities that those data structures represent. Then functions that accept such values as input can be constrained with those traits, and we can get some of the same benefits without needing all the phantoms and such?